src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotCallDangerousMethodsInDeserializationTests.cs

```diff
@@ -0,0 +1,1817 @@
+            File.WriteAllBytes(""C:\\"", bytes);
+        }
+      
+        disposed = true;
+    }
+}");
+
+            VerifyBasic(@"
+Imports System
+Imports System.IO
+Imports System.Runtime.Serialization
+
+Namespace TestNamespace
+    Class TestClass
+        Implements IDisposable
+        Private member As String
+        Protected disposed As Boolean = False
+        
+        Sub Dispose() Implements IDisposable.Dispose
+            Dim bytes(9) As Byte
+            File.WriteAllBytes(""C:\\"", bytes)
+            Dispose(True)  
+            GC.SuppressFinalize(Me) 
+        End Sub
+
+        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
+            If Not Me.disposed Then
+                If disposing Then
+                    Dim bytes(9) As Byte
+                    File.WriteAllBytes(""C:\\"", bytes)
+                End If
+            End If
+            Me.disposed = True
+        End Sub
+    End Class
+End Namespace");
+        }
+
+        [Fact]
+        public void TestDisposeNotImplementIDisposableNoDiagnostic()
+        {
+            VerifyCSharp(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization;
+
+public class TestClass
+{
+    private string member;
+    bool disposed = false;
+
+    public void Dispose()
+    {
+        byte[] bytes = new byte[] {0x20, 0x20, 0x20};
+        File.WriteAllBytes(""C:\\"", bytes);
+        Dispose(true);
+        GC.SuppressFinalize(this);           
+    }
+    
+    protected virtual void Dispose(bool disposing)
+    {
+        if (disposed)
+        {
+            return; 
+        }
+      
+        if (disposing) 
+        {
+            byte[] bytes = new byte[] {0x20, 0x20, 0x20};
+            File.WriteAllBytes(""C:\\"", bytes);
+        }
+      
+        disposed = true;
+    }
+}");
+
+            VerifyBasic(@"
+Imports System
+Imports System.IO
+Imports System.Runtime.Serialization
+
+Namespace TestNamespace
+    Class TestClass
+        Private member As String
+        Protected disposed As Boolean = False
+        
+        Sub Dispose()
+            Dim bytes(9) As Byte
+            File.WriteAllBytes(""C:\\"", bytes)
+            Dispose(True)  
+            GC.SuppressFinalize(Me) 
+        End Sub
+
+        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
+            If Not Me.disposed Then
+                If disposing Then
+                    Dim bytes(9) As Byte
+                    File.WriteAllBytes(""C:\\"", bytes)
+                End If
+            End If
+            Me.disposed = True
+        End Sub
+    End Class
+End Namespace");
+        }
+
+        protected override DiagnosticAnalyzer GetBasicDiagnosticAnalyzer()
+        {
+            return new DoNotCallDangerousMethodsInDeserialization();
+        }
+
+        protected override DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
+        {
+            return new DoNotCallDangerousMethodsInDeserialization();
+        }
+    }
+}
```

Comment:
Test cases involving a serializable base and a serializable derived class would be neat.

Are there any test cases with recursive methods?

Reply:
TestOnDeserializedMultiMediateInvocationsDiagnostic tests the recursive invocation case. And I'll add the derive case as you recommended.